### PR TITLE
 Port remaining documented C# code style rules to new format

### PR DESCRIPTION
**Pull request by mavasani**
[Preview link](https://review.docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/?branch=pr-en-us-20892)

### CODE REVIEW

**Comment by mavasani** [comment]
There seem to be quite a few undocumented IDE rules. I will work on it early next week, and once that is done, we can hook up IDE analyzers to directly link to these help pages.

File: docs/fundamentals/code-analysis/style-rules/using-directive-preferences.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+description: "Learn about code analysis language rules for 'using' directive preferences"
+ms.date: 09/30/2020
+ms.topic: reference
+author: gewarren
+ms.author: gewarren
+dev_langs:
+- CSharp
+- VB
+---
+# 'using' directive preferences
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/using-directive-preferences.md:12]: For the language rules, it looks like we don't have an overview page when there's only a single rule in the section. I think that makes more sense that having this sort of "extra" file that doesn't add much value.

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+- CSharp
+- VB
+---
+# Expression-bodied-members
+
+## Overview
+
+The style rules in this section concern the use of [expression-bodied members](/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members) when the logic consists of a single expression.
+
+- [Use expression body for constructors](ide0021.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:18]:
```suggestion
- [Use expression body for constructors (IDE0021)](ide0021.md)
```

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+- [Use expression body for methods](ide0022.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:19]:
```suggestion
- [Use expression body for methods (IDE0022)](ide0022.md)
```

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+- [Use expression body for operators](ide0023-ide0024.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:20]:
```suggestion
- [Use expression body for operators (IDE0023 and IDE0024)](ide0023-ide0024.md)
```

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+- [Use expression body for properties](ide0025.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:21]:
```suggestion
- [Use expression body for properties (IDE0025)](ide0025.md)
```

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+- [Use expression body for indexers](ide0026.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:22]:
```suggestion
- [Use expression body for indexers (IDE0026)](ide0026.md)
```

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+- [Use expression body for accessors](ide0027.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:23]:
```suggestion
- [Use expression body for accessors (IDE0027)](ide0027.md)
```

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+- [Use expression body for lambda exressions](ide0053.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:24]:
```suggestion
- [Use expression body for lambda exressions (IDE0053)](ide0053.md)
```

File: docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+- [Use expression body for local functions](ide0061.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/expression-bodied-members.md:25]:
```suggestion
- [Use expression body for local functions (IDE0061)](ide0061.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0007-ide0008.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+|-|-|
+| **Rule ID** | IDE0007 and IDE0008 |
+| **Title** | IDE0007: Use 'var' instead of explicit type<br/> IDE0008: Use explicit type instead of 'var' |
+| **Category** | Style |
+| **Subcategory** | Language rules |
+| **Applicable languages** | C# |
+
+## Overview
+
+The style rules in this section concern the use of the [var](/dotnet/csharp/language-reference/keywords/var) keyword versus an explicit type in a variable declaration. This rule can be applied separately to built-in types, when the type is apparent, and elsewhere.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0007-ide0008.md:35]:
```suggestion
The style rules in this section concern the use of the [var](/dotnet/csharp/language-reference/keywords/var) keyword versus an explicit type in a variable declaration. This rule can be applied separately to built-in types, places where the type is apparent, and elsewhere.
```

File: docs/fundamentals/code-analysis/style-rules/ide0007-ide0008.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+// csharp_style_var_elsewhere = true
+var f = this.Init();
+
+// csharp_style_var_elsewhere = false
+bool f = this.Init();
+```
+
+## See also
+
+- [Code style language rules](language-rules.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0007-ide0008.md:93]:
```suggestion
- [var keyword](/dotnet/csharp/language-reference/keywords/var)
- [Code style language rules](language-rules.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0016.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+this.s = s ?? throw new ArgumentNullException(nameof(s));
+
+// csharp_style_throw_expression = false
+if (s == null) { throw new ArgumentNullException(nameof(s)); }
+this.s = s;
+```
+
+## See also
+
+- [Null-checking preferences](null-checking-preferences.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0016.md:52]:
```suggestion
- [throw expressions](../../../csharp/language-reference/keywords/throw.md#the-throw-expression)
- [Null-checking preferences](null-checking-preferences.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0020.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_style_pattern_matching_over_is_with_cast_check = true
+if (o is int i) {...}
+
+// csharp_style_pattern_matching_over_is_with_cast_check = false
+if (o is int) {var i = (int)o; ... }
+```
+
+## See also
+
+- [Pattern matching preferences](pattern-matching-preferences.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0020.md:51]:
```suggestion
- [Pattern matching in C#](../../../csharp/pattern-matching.md)
- [Pattern matching preferences](pattern-matching-preferences.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0021.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_style_expression_bodied_constructors = true
+public Customer(int age) => Age = age;
+
+// csharp_style_expression_bodied_constructors = false
+public Customer(int age) { Age = age; }
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0021.md:51]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0022.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_style_expression_bodied_methods = true
+public int GetAge() => this.Age;
+
+// csharp_style_expression_bodied_methods = false
+public int GetAge() { return this.Age; }
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0022.md:51]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0023-ide0024.md [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    => new ComplexNumber(c1.Real + c2.Real, c1.Imaginary + c2.Imaginary);
+
+// csharp_style_expression_bodied_operators = false
+public static ComplexNumber operator + (ComplexNumber c1, ComplexNumber c2)
+{ return new ComplexNumber(c1.Real + c2.Real, c1.Imaginary + c2.Imaginary); }
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0023-ide0024.md:55]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0025.md [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_style_expression_bodied_properties = true
+public int Age => _age;
+
+// csharp_style_expression_bodied_properties = false
+public int Age { get { return _age; }}
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0025.md:51]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0026.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_style_expression_bodied_indexers = true
+public T this[int i] => _values[i];
+
+// csharp_style_expression_bodied_indexers = false
+public T this[int i] { get { return _values[i]; } }
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0026.md:51]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0027.md [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_style_expression_bodied_accessors = true
+public int Age { get => _age; set => _age = value; }
+
+// csharp_style_expression_bodied_accessors = false
+public int Age { get { return _age; } set { _age = value; } }
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0027.md:51]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0053.md [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_style_expression_bodied_lambdas = true
+Func<int, int> square = x => x * x;
+
+// csharp_style_expression_bodied_lambdas = false
+Func<int, int> square = x => { return x * x; };
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/style-rules/ide0053.md:51]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0016.md [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+|-|-|
+| **Rule ID** | IDE0016 |
+| **Title** | Use throw expression |
+| **Category** | Style |
+| **Subcategory** | Language rules (null-checking preferences) |
+| **Applicable languages** | C# 7.0+ |
+
+## Overview
+
+This style rule concerns with preference of use of `throw` expressions instead of `throw` statements.
```
[the remark this answers is not in the document]
Comment (mavasani) [reply docs/fundamentals/code-analysis/style-rules/ide0016.md:29]: I'll make this grammar change in all files.

File: docs/fundamentals/code-analysis/style-rules/ide0019.md [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+if (o is string s) {...}
+
+// csharp_style_pattern_matching_over_as_with_null_check = false
+var s = o as string;
+if (s != null) {...}
+```
+
+## See also
+
+- [Pattern matching preferences](pattern-matching-preferences.md)
```
Comment (mavasani) [inline docs/fundamentals/code-analysis/style-rules/ide0019.md:52]:
```suggestion
- [Pattern matching in C#](../../../csharp/pattern-matching.md)
- [Pattern matching preferences](pattern-matching-preferences.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0034.md [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+// csharp_prefer_simple_default_expression = true
+void DoWork(CancellationToken cancellationToken = default) { ... }
+
+// csharp_prefer_simple_default_expression = false
+void DoWork(CancellationToken cancellationToken = default(CancellationToken)) { ... }
+```
+
+## See also
+
+- [Expression-level preferences](expression-level-preferences.md)
```
Comment (mavasani) [inline docs/fundamentals/code-analysis/style-rules/ide0034.md:51]:
```suggestion
- [`default` literal](../../../csharp/language-reference/operators/default#default-literal)
- [Expression-level preferences](expression-level-preferences.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0056.md [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+var index = names[^1];
+
+// csharp_style_prefer_index_operator = false
+string[] names = { "Archimedes", "Pythagoras", "Euclid" };
+var index = names[names.Length - 1];
+```
+
+## See also
+
+- [Expression-level preferences](expression-level-preferences.md)
```
Comment (mavasani) [inline docs/fundamentals/code-analysis/style-rules/ide0056.md:53]:
```suggestion
- [index-from-end operator](../../../csharp/language-reference/operators/member-access-operators.md#index-from-end-operator)
- [Expression-level preferences](expression-level-preferences.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0057.md [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+var sub = sentence[0..^4];
+
+// csharp_style_prefer_range_operator = false
+string sentence = "the quick brown fox";
+var sub = sentence.Substring(0, sentence.Length - 4);
+```
+
+## See also
+
+- [Expression-level preferences](expression-level-preferences.md)
```
Comment (mavasani) [inline docs/fundamentals/code-analysis/style-rules/ide0057.md:53]:
```suggestion
- [range operator](../../../csharp/language-reference/operators/member-access-operators.md#range-operator-)
- [Expression-level preferences](expression-level-preferences.md)
```

File: docs/fundamentals/code-analysis/style-rules/ide0061.md [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+    void Hello()
+    {
+        Console.WriteLine("Hello");
+    }
+}
+```
+
+## See also
+
+- [Code style rules for expression-bodied members](expression-bodied-members.md)
```
Comment (mavasani) [inline docs/fundamentals/code-analysis/style-rules/ide0061.md:62]:
```suggestion
- [Expression-bodied members](../../../csharp/programming-guide/statements-expressions-operators/expression-bodied-members.md)
- [Code style rules for expression-bodied members](expression-bodied-members.md)
```

File: docs/fundamentals/code-analysis/style-rules/using-directive-preferences.md (hunk 1 above)
Comment (mavasani) [reply docs/fundamentals/code-analysis/style-rules/using-directive-preferences.md:12]: Sure, let me delete this. We can introduce it if we end up adding another rule to this option group.

File: docs/fundamentals/code-analysis/style-rules/ide0007-ide0008.md [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+title: "IDE0007 and IDE0008: 'var' preferences"
+description: "Learn about code analysis rules IDE0007 and IDE0008: 'var' preferences"
+ms.date: 09/30/2020
+ms.topic: reference
+f1_keywords:
+- IDE0007
+- IDE0008
+- csharp_style_var_for_built_in_types
+- csharp_style_var_when_type_is_apparent
+- csharp_style_var_elsewhere
```
Comment (Youssef1313) [inline docs/fundamentals/code-analysis/style-rules/ide0007-ide0008.md:11]: Does Roslyn support these f1_keywords?

**Comment by mavasani** [comment]
@gewarren Any further feedback on this one?

**Review by gewarren — APPROVED** [review]
LGTM

**Comment by mavasani** [comment]
@gewarren I merged the PR after merge conflict resolution as it showed all green, but seems the build had warnings: https://opbuildstorageprod.blob.core.windows.net/report/2020%5C10%5C6%5Ca3bda507-3390-de91-8f7f-26f90f4e5fc8%5CPullRequest%5C202010061624020855-20892%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=8YLPcRd32ZIPITIj6KcKRj%2FWyILlz2EaqS%2F661PNGHI%3D&st=2020-10-06T16%3A20%3A45Z&se=2020-11-06T16%3A25%3A45Z&sp=r

```
.openpublishing.redirection.json 

Line 2619: [Warning] The '/dotnet/csharp/whats-new/csharp-7' appears twice or more in the redirection mappings.
Line 2624: [Warning] The '/dotnet/csharp/whats-new/csharp-7' appears twice or more in the redirection mappings.
Line 2629: [Warning] The '/dotnet/csharp/whats-new/csharp-7' appears twice or more in the redirection mappings
```

Do you want me to send a follow-up PR to fix these?

**Comment by gewarren** [comment]
@mavasani No worries, I will take care of it. Thanks for letting me know!

**Pull request merged**